### PR TITLE
[RBAC]: Frontend Part II

### PR DESCRIPTION
**Pull request by joeyorlando**
**What this PR does**:
This PR refactors `grafana-plugin/src/state/userAction.ts` into `grafana-plugin/src/utils/authorization`. Previously this was hardcoded to a list of strings that were being returned by the OnCall API. Now the values here correspond to the new permissions defined in `plugin.json` (in #685). Each `UserAction` has an RBAC permission string, and a "minimum fallback role" in cases where RBAC is not supported.

**Which issue(s) this PR fixes**:
`grafana/oncall-private` [#1381](https://github.com/grafana/oncall-private/issues/1381)

**Checklist**
- [x] Tests updated
- [ ] Documentation added (done in a separate PR)
- [ ] `CHANGELOG.md` updated (done in a separate PR)

RBAC PRs (once approved, all of these PRs will be merged into the `jorlando/rbac` branch and that branch will then be merged to `dev`):
- #685
- #634 
- #732 
- #687
- #699 
- #686

### CODE REVIEW

File: grafana-plugin/src/GrafanaPluginRootPage.tsx [hunk 1]
```diff
@@ -150,12 +151,20 @@ export const Root = observer((props: AppRootProps) => {
...
   return (
     <DefaultPageLayout {...props}>
       <GrafanaTeamSelect currentPage={page} />
-      <Page {...props} path={pathWithoutLeadingSlash} />
+      {userHasAccess ? (
+        <PageComponent {...props} path={pathWithoutLeadingSlash} />
+      ) : (
+        <>
+          {/* TODO: what should be the behavior here? a redirect? if so, to what page? or maybe an unauthorized component? */}
+          <p>You don't have access!</p>
```
Comment (joeyorlando) [inline grafana-plugin/src/GrafanaPluginRootPage.tsx:165]: my thoughts here are that we could either redirect the user to the first page that they have access to, or if they don't have access to any, redirect them to the main grafana dashboard page?

File: grafana-plugin/src/state/plugin.ts [hunk 2]
```diff
@@ -12,6 +12,7 @@ export async function createGrafanaToken() {
 
   return await getBackendSrv().post('/api/auth/keys', {
     name: 'OnCall',
+    // TODO: what to do about this role attribute?
     role: 'Admin',
```
Comment (joeyorlando) [inline grafana-plugin/src/state/plugin.ts:16]: not sure what we should do here?

File: grafana-plugin/src/state/rootBaseStore.ts [hunk 3]
```diff
@@ -239,7 +234,9 @@ export class RootBaseStore {
   }
 
   isUserActionAllowed(action: UserAction) {
-    return this.userStore.currentUser && this.userStore.currentUser.permissions.includes(action);
+    // https://github.com/grafana/grafana/blob/main/public/app/core/services/context_srv.ts#L165-L170
+    // TODO: what to assign fallback (2nd argument) to?
+    return contextSrv.hasAccess(action, true);
```
Comment (joeyorlando) [inline grafana-plugin/src/state/rootBaseStore.ts:239]: this is important

File: grafana-plugin/src/state/rootBaseStore.ts [hunk 4]
```diff
@@ -228,7 +222,8 @@ export class RootBaseStore {
         this.signupAllowedForPlugin = false;
         return;
       }
-      const userRole = await this.getUserRole();
+      // TODO: what to do here?!?!
+      const userRole = 'Admin';
       if (userRole !== 'Admin') {
         this.correctRoleForInstallation = false;
```
Comment (joeyorlando) [inline grafana-plugin/src/state/rootBaseStore.ts:228]: not sure what this should get refactored to?

File: grafana-plugin/src/plugin.json [hunk 5]
```diff
@@ -40,63 +40,63 @@
       "type": "page",
```
Comment (joeyorlando) [inline grafana-plugin/src/plugin.json:40]: @gamab would be good to get your eyes on this file

File: grafana-plugin/src/GrafanaPluginRootPage.tsx (hunk 1 above)
Comment (maskin25) [inline grafana-plugin/src/GrafanaPluginRootPage.tsx:165]: agree, some additional info, links can be useful here, I suppose @raphael-batte can help

File: grafana-plugin/src/plugin.json [hunk 6]
```diff
@@ -192,6 +192,457 @@
...
+  "roles": [
+    {
+      "role": {
+        "version": 1,
+        "name": "plugins.app:grafana-oncall-app:admin",
+        "displayName": "Admin",
+        "description": "Read/write access to everything in OnCall",
+        "group": "OnCall",
+        "permissions": [
+          { "action": "plugins.app:access", "scope": "plugins.app:id:grafana-oncall-app" },
```
Comment (joeyorlando) [inline grafana-plugin/src/plugin.json:204]: After discussing w/ Matvey, thinking of adding a  "OnCaller" role:
- view permissions to everything
- edit permissions to alert groups
- (maybe) edit permissions to schedules

File: grafana-plugin/src/plugin.json [hunk 7]
```diff
@@ -192,6 +192,460 @@
...
+          { "action": "grafana-oncall-app.alert-groups:read" }
+        ],
+        "hidden": false
+      },
+      "grants": []
+    },
+    {
+      "role": {
+        "name": "plugins.app:grafana-oncall-app:alert-groups-editor",
+        "displayName": "Alert Groups Edtiro",
```
Comment (joeyorlando) [inline grafana-plugin/src/plugin.json:330]:
```suggestion
        "displayName": "Alert Groups Editor",
```

File: grafana-plugin/src/state/rootBaseStore.ts (hunk 3 above)
Comment (joeyorlando) [reply grafana-plugin/src/state/rootBaseStore.ts:237]: @gamab how're you currently handling the `fallback` arg to `contextSrv.hasAccess`?

File: grafana-plugin/src/GrafanaPluginRootPage.tsx (hunk 1 above)
Comment (joeyorlando) [reply grafana-plugin/src/GrafanaPluginRootPage.tsx:165]: @raphael-batte the current UI would look like this. Would you prefer that users get redirect to some page, or that they see some unauthorized message/component?
Screen Shot 2022-10-24 at 15 05 34

File: grafana-plugin/src/state/rootBaseStore.ts (hunk 3 above)
Comment (gamab) [reply grafana-plugin/src/state/rootBaseStore.ts:237]: The fallback on our side is the legacy access-control check, it is enforce should RBAC be disabled. So if you used to protect the view to admins only, you'd have an `isAdmin` boolean there ( `contextSrv.hasRole("Admin")`)
If we say that OnCall is going to use RBAC only - which I think is valid - then use `contextSrv.hasPermission`

File: grafana-plugin/src/pages/schedules/Schedules.tsx [hunk 8]
```diff
@@ -530,6 +530,7 @@ const Event = ({ event }: EventProps) => {
                       (check if {event.missing_users[0].includes(',') ? 'some of these users -' : 'user -'}{' '}
                       <Text type="secondary">"{event.missing_users[0]}"</Text>{' '}
                       {event.missing_users[0].includes(',') ? 'are' : 'is'} existing in OnCall or{' '}
+                      {/* TODO: should probably update this message? */}
```
Comment (joeyorlando) [inline grafana-plugin/src/pages/schedules/Schedules.tsx:533]: TODO: update wording here

File: grafana-plugin/src/GrafanaPluginRootPage.tsx (hunk 1 above)
Comment (raphael-batte) [reply grafana-plugin/src/GrafanaPluginRootPage.tsx]: looks nice!
one little thing - lets correct color for 403
basic Grafana orange will be great here - rgb(255, 120, 10)

Comment (joeyorlando) [reply grafana-plugin/src/GrafanaPluginRootPage.tsx]: @raphael-batte is it okay if I create a separate issue to update this?

Currently this is using the same color as a bunch of other places in the UI (#f8d06b for dark theme, #8a6c00 for light).

Comment (joeyorlando) [reply grafana-plugin/src/GrafanaPluginRootPage.tsx]: @raphael-batte I created #768 to make sure we don't forget about this

**Pull request merged**